File: Latihan Kelas/KendaraanInheritance/KoleksiKendaraan.cpp
#include <iostream>
#include "Kendaraan.hpp"

class KoleksiKendaraan {
    private:
        int nEff;
        int size;
        Kendaraan** arr;
    public:
        KoleksiKendaraan(): nEff(0), size(100), arr(new Kendaraan*[size]) {};
        KoleksiKendaraan(int size): nEff(0), size(size), arr(new Kendaraan*[size]) {};
        KoleksiKendaraan(const KoleksiKendaraan& k) {
            nEff = k.nEff;
            for (int i = 0; i < nEff; i++) {
                arr[i] = k.arr[i];
            }
        }
        ~KoleksiKendaraan() {
            delete[] arr;
        }
        
        void printAll() {
            for (int i = 0; i < nEff; i++) {
                arr[i]->printInfo();
                cout << endl;
            }
        }
 
        KoleksiKendaraan& operator<<(Kendaraan* a) {
            if (nEff + 1 < size){
                arr[nEff] = a;
                nEff++;
            }
            return *this;
        }
        KoleksiKendaraan& operator<<(KoleksiKendaraan a) {
            if (nEff + a.nEff < size){
                for (int i = 0; i < a.nEff; i++) {
                    arr[nEff] = a.arr[i];
                    nEff++;
                }
            } else {
                for (int i = 0; i < size - nEff; i++){
                    arr[nEff] = a.arr[i];
                    nEff++;
                }
            }
            return *this;
        }
};

File: Latihan Kelas/KendaraanInheritance - Jawwa/KoleksiKendaraan.java
public class KoleksiKendaraan {
    private int nEff;
    private int size;
    private Kendaraan[] arr;

    public KoleksiKendaraan() {
        this.nEff = 0;
        this.size = 10;
        this.arr = new Kendaraan[size];
    }

    public KoleksiKendaraan(int size){
        this.nEff = 0;
        this.size = size;
        this.arr = new Kendaraan[size];
    }

    public void printAll(){
        for (int i = 0; i < this.nEff; i++){
            this.arr[i].printInfo();
        }
    }

    public void add(Kendaraan k){
        if (this.nEff + 1 < size){
            arr[nEff] = k;
            nEff++;
        } else {
            System.out.println("Koleksi penuh");
        }
    }

    public void add(KoleksiKendaraan kk){
        if (this.nEff + kk.nEff < size){
            for (int i = 0; i < kk.nEff; i++){
                this.arr[this.nEff] = kk.arr[i];
                this.nEff++;
            }
        } else {
            System.out.println("Ukuran tidak muat untuk menambahkan koleksi kendaraan baru!");
        }
    }
}
